### PR TITLE
Add realworld go sample

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatoliy Bazko <abazko@redhat.com>

### What does this PR do?
It adds real-world-go-sample
https://github.com/xesina/golang-echo-realworld-example-app

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13803

### Related PRs
https://github.com/eclipse/che-plugin-registry/pull/224
https://github.com/eclipse/che-theia/pull/443

### How to test
1. replace plugin id to with the reference
```
id: ms-vscode/go/latest
```
-> 
```
 reference: >-
      https://raw.githubusercontent.com/eclipse/che-plugin-registry/11ea1728310fd8ed6b26e904d4652d7aa14f166b/v3/plugins/ms-vscode/go/0.11.4/meta.yaml
```

2. Build `docker.io/eclipse/che-remote-plugin-go-1.12.9:next` (see https://github.com/eclipse/che-theia/pull/443)
or retag it from `docker.io/abazko/che-remote-plugin-go-1.12.9:next`

### CODE REVIEW

**Comment by tolusha** [comment]
I've added `How to test` section

File: devfiles/go-web-sample/devfile.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+apiVersion: 1.0.0
+metadata:
+  generateName: golang-web-
+projects:
+-
+  name: example
+  source:
+    type: git
+    location: https://github.com/xesina/golang-echo-realworld-example-app
+  clonePath: src/github.com/xesina/golang-echo-realworld-example-app
```
Comment (nickboldt) [inline devfiles/go-web-sample/devfile.yaml]: why not clone this into a shorter folder path, like just `golang-echo-realworld-example-app`?

File: devfiles/go-web-sample/devfile.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  memoryLimit: 512Mi
+  mountSources: true
+commands:
+-
+  name: install dependencies
+  actions:
+  - type: exec
+    component: go-cli
+    command: go mod download
+    workdir: ${CHE_PROJECTS_ROOT}/src/github.com/xesina/golang-echo-realworld-example-app
```
Comment (nickboldt) [inline devfiles/go-web-sample/devfile.yaml]: same comment as above. Use shorter path: `${CHE_PROJECTS_ROOT}/golang-echo-realworld-example-app` ?

File: devfiles/go/devfile.yaml [hunk 3]
```diff
@@ -18,7 +18,7 @@ components:
 -
   type: dockerimage
   # this version is used in the plugin
-  image: quay.io/eclipse/che-golang-1.10:nightly
+  image: quay.io/eclipse/che-golang-1.12:nightly
```
Comment (nickboldt) [inline devfiles/go/devfile.yaml]: do we need to support the go 1.10 stuff too? or just drop it in favour of the latest?

Comment (nickboldt) [reply devfiles/go/devfile.yaml]: see also https://github.com/eclipse/che-plugin-registry/pull/224

File: devfiles/go-web-sample/devfile.yaml (hunk 1 above)
Comment (amisevsk) [reply devfiles/go-web-sample/devfile.yaml]: My guess would be `$GOPATH` compatibility issues. It's a similar case for our current Go sample.

File: devfiles/go-web-sample/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+displayName: Go Real World Application
+description: Go Stack with Go 1.12.9 and Real Wold Application
```
Comment (benoitf) [inline devfiles/go-web-sample/meta.yaml]: just to inform there is a typo on WoRld
```suggestion
description: Go Stack with Go 1.12.9 and Real World Application
```

File: devfiles/go-web-sample/devfile.yaml (hunk 1 above)
Comment (skabashnyuk) [reply devfiles/go-web-sample/devfile.yaml]: https://github.com/xesina/golang-echo-realworld-example-app uses go modules https://blog.golang.org/using-go-modules. Are we sure we need to use $GOPATH?

File: devfiles/go/devfile.yaml (hunk 3 above)
Comment (tolusha) [reply devfiles/go/devfile.yaml]: @nickboldt @amisevsk 
I have no idea

File: devfiles/go-web-sample/devfile.yaml (hunk 1 above)
Comment (tolusha) [reply devfiles/go-web-sample/devfile.yaml]: yes, it works.

File: devfiles/go-web-sample/devfile.yaml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  clonePath: src/github.com/xesina/golang-echo-realworld-example-app
+components:
+-
+  type: chePlugin
+  id: ms-vscode/go/latest
+  alias: go-plugin
+  memoryLimit: 512Mi
+-
+  type: dockerimage
+  # this version is used in the plugin
```
[the remark this answers is not in the document]
Comment (tolusha) [reply devfiles/go-web-sample/devfile.yaml]: I don't think so.

**Comment by tolusha** [comment]
@amisevsk  @l0rd @tsmaeder 
> I'm generally against adding pure backend samples to the devfiles registry as they provide a rather poor user experience (see the nodejs-mongo sample we're currently working on fixing).

It was decided here https://github.com/eclipse/che/issues/13529
A lot of work has been done since then. Does it mean it was for nothing?

File: devfiles/go/devfile.yaml (hunk 3 above)
Comment (amisevsk) [reply devfiles/go/devfile.yaml]: I'm currently looking to deprecate 1.10 if possible, so if the devfiles work with 1.12, that makes my life easier.

See: https://github.com/eclipse/che/issues/14265, https://github.com/eclipse/che-devfile-registry/pull/79#discussion_r313518538

**Comment by amisevsk** [comment]
> A lot of work has been done since then. Does it mean it was for nothing?

AFAICT the work that was done was very important for running Go on Che in the general case (if you're referring to issue https://github.com/eclipse/che/issues/13803), so the work is not for nothing. The issue I have with applications that are only backend with no frontend is that we run into https://github.com/eclipse/che/issues/14392 (which runs the same realworld backend app).

**Comment by tolusha** [comment]
I observed several issues:

1. Should be fixed by: https://github.com/eclipse/che/issues/13614
```
Uncaught Error: 'file:///go/pkg/mod/github.com/jinzhu/gorm%40v1.9.8/main.go' has not been found.

Error: Request 'resolveContent' failed
    at Proxy.<anonymous> (https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1786120)
    at e.<anonymous> (https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1982608)
    at https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1981179
    at Object.next (https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1981284)
    at https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1980197
    at new Promise (<anonymous>)
    at T1Xs.a (https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1979974)
    at e.readContents (https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1982467)
    at new e (https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:4362797)
    at e.<anonymous> (https://static.developers.redhat.com/che/theia_artifacts/theia.d799fbdd096d798c68c8.js:1:1469480)
```

2.  
```
Uncaught Error: Command failed: /go/bin/godef -t -i -f /projects/golang-web-app/store/user.go -o 177
Error: Command failed: /go/bin/godef -t -i -f /projects/golang-web-app/store/user.go -o 177
```

**Comment by tolusha** [comment]
@amisevsk 
The PR is updated and is ready to test.

**Comment by tolusha** [comment]
> Additionally: the golang plugin seems to be unable to handle methods on structs that are defined in other files (see article.go, which defines methods on handler). This is not an issue for this PR, though.

Actually it works. The trick is to set the project as a workspace root. But there is no way to do it automatically, that's way this issue became important https://github.com/eclipse/che/issues/13080

screencast-routeiu9ly0wn-che 192 168 99 100 nip io-2019 11 06-09_54_04

**Comment by tolusha** [comment]
@amisevsk 
From my point of view this issue [1] should be fixed before merging.
As far as I understood, we have to replace the project with one having a proper license, correct?

[1] https://github.com/eclipse/che/issues/13080

**Comment by tsmaeder** [comment]
@tolusha any reason we're using the the version from "xesina" instead of the version it was forked from (which is MIT licensed)?

**Comment by tsmaeder** [comment]
> As far as I understood, we have to replace the project with one having a proper license, correct?

Strictly speaking, no we don't: we're not distributing the sample, just pointing to some code on github. However, if we want to distribute the sample (in a airgapped version of CRW), we'll have to have the licensing in order. 
That said, as an open source project, we should use examples that follow good practices for licensing, so yes, we should find an example that is properly licensed.

**Comment by tolusha** [comment]
@tsmaeder 
> any reason we're using the the version from "xesina" instead of the version it was forked from (which is MIT licensed)?

Which one? Probably I am not aware of.

**Comment by tsmaeder** [comment]
The repo is a github fork of another github project (which has a license)

**Comment by tolusha** [comment]
see comment https://github.com/eclipse/che/issues/13803#issuecomment-561087535

**Pull request closed without merging**